### DISCUSSION TITLE
Issue with Redirect Loop on 404 Page

### DISCUSSION THREAD

**Discussion by muhammad-umar-bhutta** [vercel/next.js#50732] [Help]
### Summary

Hello everyone! I'm currently facing an issue with my website where I have several pages such as page.js, layout.js, not-found.js, contact-us, about-us, and slug. The problem occurs when I manually enter a URL that doesn't exist or when there is a programmatic redirection to the 404 page. Instead of displaying the 404 page and stopping, it goes into a loop and keeps refreshing the page repeatedly. I would greatly appreciate your assistance in resolving this issue. The desired behavior is for the 404 page to be displayed only once when a non-existent URL is accessed, without going into a loop. Thank you in advance for your help.

Here's the code snippet that is causing the issue:

### Additional information

```js
import Link from 'next/link';

export default function NotFound() {
  return (
    <>
      <main className="bg-gray-50 h-[70vh] flex justify-center items-center">
        <div className="container px-4 py-8">
          <div className="max-w-2xl mx-auto text-center">
            <h1 className="text-5xl font-bold font-heading">404</h1>
            <p className="text-2xl font-medium mb-8">Page Not Found</p>
            <p className="text-lg mb-4">The page you are looking for could not be found.</p>
            <p className="text-lg mb-4">Please check the URL or return to the <Link href="/"><span className="text-blue-600">homepage</span></Link>.</p>
          </div>
        </div>
      </main>
    </>
  );
}
```

**Comment by devjiwonchoi** [comment]
Could you provide your folder structure and the page calling the `notFound` function?

**Reply by muhammad-umar-bhutta** [reply]
I encountered an error even after changing the name of the page to "notFound." Here's the updated code snippet:
import Link from 'next/link';

export default function notFound() {
  return (
    <>
      <div className="bg-gray-50 h-[70vh] flex justify-center items-center">
        <div className="container px-4 py-8">
          <div className="max-w-2xl mx-auto text-center">
            <h1 className="text-5xl font-bold font-heading">404</h1>
            <p className="text-2xl font-medium mb-8">Page Not Found</p>
            <p className="text-lg mb-4">The page you are looking for could not be found.</p>
            <p className="text-lg mb-4">Please check the URL or return to the <Link href="/"><span className="text-blue-600">homepage</span></Link>.</p>
          </div>
        </div>
      </div>
    </>
  );
}

Furthermore, The project structure consists of the following directories and files:

Root Directory
layout.js
page.js
loading.js
not-found.js
components (contains various components)
slug-pages (includes slug pages)
marketing (contains about-us and contact-us folders, which have their respective pages)

**Reply by devjiwonchoi** [reply]
This issue has been out for a long time, see #10024 and #50429.

Note that it happens only in `development`, not in `production`.

I'm assuming that it is related to `Fast Refresh` which is used in the development stage for Next.js since `Page Router`,

> In the development stage, Next.js optimizes for the developer and their experience building the application. It comes with features that aim to improve the Developer Experience such the [TypeScript](https://nextjs.org/docs/basic-features/typescript) and [ESLint integration](https://nextjs.org/docs/basic-features/eslint), [Fast Refresh](https://nextjs.org/docs/basic-features/fast-refresh), and more.

but nothing has been clearly found out yet.

**Reply by muhammad-umar-bhutta** [reply]
As a developer, having a conducive development environment is crucial for efficient workflow. Currently, I'm facing some challenges when it comes to handling breakpoints in certain pages, specifically in the slug functionality. I need to implement different actions based on whether a slug is found or not. However, when rendering to the 404 page, it continuously reloads during development, negatively impacting the developer experience.

I appreciate the valuable features provided by Next.js, which have been instrumental in saving time and enhancing the user experience. Overall, I'm quite fond of Next.js. However, there are instances when both the client and I require fast execution, and it would be greatly beneficial if these issues could be resolved promptly.

Additionally, I would appreciate more details on implementing the top-loading bar in my existing application. In previous versions, it worked well by utilizing the router.events properties. However, with the introduction of Next.js 13 and the next/navigation feature, I'm facing difficulties executing it properly. I kindly request your assistance in resolving this issue as soon as possible.

Thank you for your support and prompt attention to these matters.

**Comment by TangoPJ** [comment]
I have the same problem but in the production mode only, in the dev mode all ok. I created a multi language landing page with this example - https://github.com/i18next/next-i18next/blob/master/examples/ssg/lib/redirect.js
